Hermes/appscale/hermes/stats/pubsub_base.py

```diff
@@ -0,0 +1,147 @@
+""" Base classes for building stats publishers and subscribers """
+import logging
+import sys
+
+from tornado.ioloop import PeriodicCallback, IOLoop
+
+
+class SubscriberIsAlreadyRegistered(Exception):
+  """ Raised when you register single subscriber twice on one publisher """
+  pass
+
+
+class SubscriberIsNotRegistered(Exception):
+  """ Raised when you unregister subscriber which is not registered yet """
+  pass
+
+
+class StatsSource(object):
+  """
+  Base class for producers of any kind of stats.
+  Subclasses should implement method get_current()
+  """
+  @property
+  def stats_name(self):
+    return self.__class__.__name__
+
+  def __repr__(self):
+    return self.stats_name
+
+  def get_current(self):
+    """ Returns: current value of specific kind of stats """
+    raise NotImplementedError()
+
+
+class AsyncStatsSource(object):
+  """
+  Base class for asynchronous producers of any kind of stats.
+  Subclasses should implement get_current_async()
+  """
+  @property
+  def stats_name(self):
+    return self.__class__.__name__
+
+  def __repr__(self):
+    return self.stats_name
+
+  def get_current_async(self):
+    """ Returns: Future wrapper for current value of specific kind of stats """
+    raise NotImplementedError()
+
+
+class StatsSubscriber(object):
+  """
+  Base class for stats consumers.
+  Subclasses should implement method receive(stats)
+  """
+  @property
+  def subscriber_name(self):
+      return self.__class__.__name__
+
+  def __repr__(self):
+    return self.subscriber_name
+
+  def receive(self, stats):
+    """ Handlers produced stats entity
+    Args:
+      stats: an object containing stats of some kind (node, processes, ...)
+    """
+    raise NotImplementedError()
+
+
+class StatsPublisher(object):
```

Comment:
When I add profile log, it will appear to be "collect and cache and write to log".
When portal support is added it can become "collect and cache and write to log and send to portal.

My idea was that there are 3 consumers of stats: local cache to deal with temporary network issues, profile log and Portal.

So I still believe it can be called as it is.